sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionManagerSample.ts

```diff
@@ -0,0 +1,155 @@
+  processInitialize(sessionId: string);
+  processMessage(msg: ReceivedMessageWithLock);
+  processError(err: Error, sessionId: string);
+  processClose(reason: "error" | "idle_timeout", sessionId: string);
+}
+
+function _createIdleTimer(
+  timeoutMs: number
+): { refresh(): void; expirationPromise: Promise<void> } {
+  let resolveExpirationPromise;
+
+  const expirationPromise = new Promise<void>((res, _rej) => {
+    resolveExpirationPromise = res;
+  });
+
+  let timer = setTimeout(resolveExpirationPromise, timeoutMs);
+
+  return {
+    refresh() {
+      clearTimeout(timer);
+      timer = setTimeout(resolveExpirationPromise, timeoutMs);
+    },
+    expirationPromise
+  };
+}
+
+async function _noSessionsAvailable(err: Error) {
+  // TODO: this error should be programmatically identifable.
+  return err.message.search(/Received an incorrect sessionId 'undefined'/) !== -1;
+}
+
+async function _processNextSession(
+  serviceBusClient: ServiceBusClient,
+  handlers: SessionMessageHandlers,
+  abortSignal: AbortSignalLike
+): Promise<void> {
+  const sessionReceiver = serviceBusClient.createSessionReceiver(queueName, "peekLock", {
+    autoRenewLockDurationInMs: sessionIdleTimeoutMs
+  });
+
+  try {
+    // TODO: I just picked a call that didn't grab a message here, just to ensure a session could
+    // actually be obtained. This goes away when we have `await createSessionReceiver()`.
+    await sessionReceiver.getState();
+  } catch (err) {
+    // TODO: when this happens do I still need to close the session receiver? (again, temporary concern)
+    if (_noSessionsAvailable(err)) {
+      console.log(`INFO: no available sessions, sleeping for ${delayWhenNoSessionsAvailableMs}`);
+      await delay(delayWhenNoSessionsAvailableMs);
+      return;
+    }
+
+    throw err;
+  }
+
+  await handlers.processInitialize(sessionReceiver.sessionId);
+
+  const idleTimer = _createIdleTimer(sessionIdleTimeoutMs);
+  let threwError = false;
+
+  sessionReceiver.subscribe(
+    {
+      async processMessage(msg) {
+        idleTimer.refresh();
+        await handlers.processMessage(msg);
+      },
+      async processError(err) {
+        threwError = true;
+        await handlers.processError(err, sessionReceiver.sessionId);
+      }
+    },
+    {
+      abortSignal
+    }
+  );
+
+  await idleTimer.expirationPromise;
+
+  try {
+    await sessionReceiver.close();
+  } catch (err) {
+    await handlers.processError(err, sessionReceiver.sessionId);
+  }
+
+  await handlers.processClose(threwError ? "error" : "idle_timeout", sessionReceiver.sessionId);
+}
+
+async function _roundRobinThroughAvailableSessions(
+  handlers: SessionMessageHandlers,
+  abortSignal: AbortSignalLike
+): Promise<void> {
+  const serviceBusClient = new ServiceBusClient(serviceBusConnectionString);
+  const allPromises = [];
+
+  for (let i = 0; i < maxSessionsToProcessSimultaneously; ++i) {
+    allPromises.push(
+      new Promise(async (_res, rej) => {
```

Comment:
We can skip creating the new promise and the try/catch here if we move the while loop into `_processNextSession` right? Or create a new function 

```javascript
async function keepProcessingSessions(serviceBusClient, handlers, abortSignal) {
      while (!abortSignal.aborted) {
            await _processNextSession(serviceBusClient, handlers, abortSignal);
          }
}
```

Simplifies the for loop to 

```js
for (let i = 0; i < maxSessionsToProcessSimultaneously; ++i) {
   allPromises.push(keepProcessingSessions(serviceBusClient, handlers, abortSignal))
}

Reply:
I still kept the while loop on the outside, but I did eliminate the Promise (and also the push, which removed even more).

I like having the while loop out of the function so the function doesn't have this extra level of indentation, etc... if people want to pick through it.